Source/NSPathControl.m

```diff
@@ -0,0 +1,565 @@
+
+          [i setTitle: s];
+          [i setTarget: self];
+          [i setAction: @selector(_doMenuAction:)];
+          [i setRepresentedObject: c];
+          
+          [menu insertItem: i
+                   atIndex: 0]; 
+          RELEASE(i);
+        }
+
+      // Add separator
+      [menu insertItem: [NSMenuItem separatorItem]
+               atIndex: 0];
+
+      // Add choose menu option
+      NSMenuItem *i = [[NSMenuItem alloc] init];
+      [i setTitle: @"Choose..."];
+      [i setTarget: self];
+      [i setAction: @selector(_doChooseMenuAction:)];
+      [menu insertItem: i
+               atIndex: 0];
+      RELEASE(i);
+
+      [self setMenu: menu];
+      // RELEASE(menu);
+      
+      if (_delegate)
+        {
+          if ([(id)_delegate respondsToSelector: @selector(pathControl:willPopUpMenu:)])
+            {
+              [_delegate pathControl: self
+                       willPopUpMenu: menu];
+            }
+        }
+      
+      if ([_cell delegate])
+        {
+          if ([(id)[_cell delegate] respondsToSelector: @selector(pathCell:willPopUpMenu:)])
+            {
+              [[_cell delegate] pathCell: _cell
+                           willPopUpMenu: menu];
+            }
+        }
+
+      
+      [menu popUpMenuPositionItem: [menu itemAtIndex: 0]
+                       atLocation: NSMakePoint(0.0, 0.0)
+                           inView: self];
+    }
+  else
+    {
+      NSPathComponentCell *pcc = [_cell pathComponentCellAtPoint: [event locationInWindow]
+                                                       withFrame: [self frame]
+                                                          inView: self];
+      [_cell _setClickedPathComponentCell: pcc];
+      if ([[self cell] action])
+        {
+          [self sendAction: [[self cell] action]
+                        to: [[self cell] target]];
+        }
+    }
+}
+
+- (NSDragOperation) draggingEntered: (id<NSDraggingInfo>)sender
+{
+  if (_delegate != nil)
+    {
+      NSDragOperation dop = [_delegate pathControl: self
+                                      validateDrop: sender];
+      return dop;
+    }
+  
+  return NSDragOperationNone;
+}
+
+- (BOOL) performDragOperation: (id<NSDraggingInfo>)sender
+{
+  NSPasteboard *pb = [sender draggingPasteboard];
+  if ([[pb types] containsObject: NSFilenamesPboardType])
+    {
+      NSArray *files = [pb propertyListForType: NSFilenamesPboardType];
+      if ([files count] > 0)
+        {
+          NSString *file = [files objectAtIndex: 0];
+          NSURL *u = [NSURL URLWithString: file];
+          BOOL accept = [_delegate pathControl: self
+                                    acceptDrop: sender];
+          if (accept)
+            {
+              [self setURL: u];
+            }
+          else
+            {
+              return NO;
+            }
+        }
+    }
+  return YES;
+}
+
+- (void) setFrame: (NSRect)frame
+{
+  [super setFrame: frame];
+  [self resetCursorRects];
```

Comment:
This should not be needed as it already happens in NSView. That was the idea of renaming the method,